src/fswAlgorithms/_GeneralModuleFiles/ekfInterface.cpp

```diff
@@ -0,0 +1,379 @@
+}
+
+
+/*! Update the covariance using the Joseph form of the update
+@param Eigen::MatrixXd measMat
+@param Eigen::MatrixXd noise
+@param Eigen::MatrixXd kalmanGain
+@return void
+ */
+void EkfInterface::updateCovariance(const Eigen::MatrixXd &measMat, const Eigen::MatrixXd &noise, const Eigen::MatrixXd &kalmanGain){
+    Eigen::MatrixXd josephTransform(this->state.size(), this->state.size());
+    josephTransform = Eigen::MatrixXd::Identity(this->state.size(), this->state.size()) - kalmanGain*measMat;
+    this->covar = josephTransform*this->covar*josephTransform.transpose();
+    this->covar += kalmanGain*noise*kalmanGain.transpose();
+}
+
+
+/*! Classical Kalman Filter Update (the reference state is unchanged)
+@param Eigen::MatrixXd kalmanGain
+@param Eigen::VectorXd measurementDelta
+@param Eigen::MatrixXd measurementMatrix
+@return void
+ */
+void EkfInterface::ckfUpdate(const Eigen::MatrixXd &kalmanGain,
+                             const Eigen::VectorXd &measurementDelta,
+                             const Eigen::MatrixXd &measurementMatrix){
+
+    this->stateError = this->stateError + kalmanGain*(measurementDelta - measurementMatrix*this->stateError);
+    this->updatedWithCkf = true;
+}
+
+/*! Extended Kalman Filter Update (the reference state is updated given the state error)
+@param Eigen::MatrixXd kalmanGain
+@param Eigen::VectorXd measurementDelta
+@return void
+ */
+void EkfInterface::ekfUpdate(const Eigen::MatrixXd &kalmanGain, const Eigen::VectorXd &measurementDelta){
+
+    this->stateError = kalmanGain*measurementDelta;
+    this->state += this->stateError;
+    this->updatedWithCkf = false;
+}
+
+/*! Compute the measurement residuals at a given time.
+@param Measurement
+@return Eigen::VectorXd
+ */
+Eigen::VectorXd EkfInterface::computeResiduals(const Measurement &measurement)
+{
+    Eigen::VectorXd measurementDelta(measurement.observation - measurement.model(this->state));
+    Eigen::MatrixXd measurementMatrix = computeMeasurementMatrix(this->state);
+
+    return measurementDelta - measurementMatrix*this->stateError;
+}
+
+/*!- Order the measurements chronologically (standard sort)
+ @return void
+ */
+void EkfInterface::orderMeasurementsChronologically(){
+    std::sort(this->measurements.begin(), this->measurements.end(),
+              [](std::optional<Measurement> meas1, std::optional<Measurement> meas2){
+                    if (!meas1.has_value()){return false;}
+                    else if (!meas2.has_value()){return true;}
+                    else {return meas1.value().timeTag < meas2.value().timeTag;}
+                                  });
+}
+
+/*! Runge-Kutta 4 (RK4) function for state propagation
+    @param ODEfunction function handle that includes the equations of motion
+    @param X0 initial state
+    @param t0 initial time
+    @param dt time step
+    @return Eigen::VectorXd
+*/
+Eigen::VectorXd EkfInterface::rk4(const std::function<Eigen::VectorXd(double, Eigen::VectorXd)>& ODEfunction,
```

Comment:
Gosh we should have a more generic place for an rk4

Reply:
it's in a bunch of my TALONS code, and I leave it there pretty much on purpose as an eye-sore so that we are forced to find a better home for it.